python-package/lightgbm/basic.py

```diff
@@ -546,9 +551,8 @@ def _data_from_pandas(data, feature_name, categorical_feature, pandas_categorica
             raise ValueError("DataFrame.dtypes for data must be int, float or bool.\n"
                              "Did not expect the data types in the following fields: "
                              f"{bad_index_cols_str}")
-        data = data.values
-        if data.dtype != np.float32 and data.dtype != np.float64:
-            data = data.astype(np.float32)
+        target_dtype = _find_commmon_dtype(data.dtypes)
+        data = data.to_numpy(dtype=target_dtype, na_value=np.nan)
```

Comment:
Ah good point. The problem with these dtypes is that they use `pd.NA` as their missing value, so using `.values` with this value present returns an array with object dtype because `pd.NA` has type `NAType` and numpy can't convert it to float. I'll try to find another approach.

Reply:
Got it, `data.astype(target_dtype).values` works.